mlir/test/Transforms/test-simplify-without-canonicalize.mlir

```diff
@@ -0,0 +1,33 @@
+// RUN: mlir-opt --canonicalize='enable-patterns=AnyPattern region-simplify=aggressive' %s | FileCheck %s
+
+// Perform merge checks without performing canonicalization prior to simplification
+
+// This test should not merge ^bb2 and ^bb5, despite the fact that they are
+// identical because %4 is used outside of ^bb2.
```

Comment:
So on this example, there is no different in block merging before and after the patch is there? Like the same number of blocks are present in the output as far as I can tell?

Reply:
The number of blocks is the same before and after the patch, but the `CHECK` item fails without the patch.

Reply:
Right, my point is that the comment is misleading here.

What about this reduction of the test case:

```
module {
  func.func @nested_loop(%arg0: i32, %arg1: i32, %arg2: i32, %arg3: i32, %arg4: i32, %arg5: i1) {
  // Irreducible control-flow: enter the middle of the loop in LoopBody_entry here.
    "test.foo_br"(%arg0, %arg4)[^LoopBody_entry] : (i32, i32) -> ()

  // Loop exit condition: jump to exit or 
  ^Loop_header:  // 2 preds: ^bb2, ^bb3
    // Consumes the block arg from LoopBody_entry
    // Because of this use here, we can't merge the two blocks below.
    "test.foo_br2"(%0)[^EXIT, ^LoopBody_entry, ^LoopBody_other] : (i32) -> ()

  // LoopBody_entry is jumped in from the entry block (bb0) and Loop_header
  // It **dominates** the Loop_header.
  ^LoopBody_entry(%0: i32):  // 2 preds: ^bb0, ^Loop_header
   // CHECK: test.foo
    %1 = "test.foo"(%0) : (i32) -> i32
    cf.br ^Loop_header

  // Other block inside the loop, not dominating the header
  ^LoopBody_other(%2: i32):  // pred: ^Loop_header
   // CHECK: test.foo
    %3 = "test.foo"(%2) : (i32) -> i32
    cf.br ^Loop_header

  ^EXIT:  // pred: ^Loop_header
    return
  }
}
```


Seems to me that it's more minimal and also more clear: we actually check that the block merge does not happen here.

Reply:
Yes, that reduced test case looks reasonable to me, with the exception that - as written - it would pass without this PR because the `CHECK: test.foo` lines can accidentally match the `test.foo_br*` lines. If it's ok with you, I think we should change `test.foo` to `test.bar` so they don't accidentally match the wrong operations.

Updated in 395aeacf9ab973856279af5b0231f86f75c3cc2a.